### PR TITLE
feat(slider): add Slider component

### LINKED ISSUE #143: ### Slider

### PR DESCRIPTION
**Pull request by Ryan2128**
- [x] Component Migration

Fix #143 

But, still a feat, adapt InputNumber component, should to do after it refactored.

### CODE REVIEW

File: packages/slider/doc/basic-usage.vue [hunk 1]
```diff
@@ -0,0 +1,52 @@
+<template>
+  <div>
```
Comment (Amour1688) [inline packages/slider/doc/basic-usage.vue:2]: Fragment

File: packages/slider/src/index.vue [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+    setValues()
+  })
+
+  watch(() => props.min, () => {
+    setValues()
+  })
+
+  watch(() => props.max, () => {
+    setValues()
+  })
```
Comment (JeremyWuuuuu) [inline packages/slider/src/index.vue:372]:
```suggestion
watch(() => [props.min, props.max], () => {
  setValues()
})
```

File: packages/slider/src/index.vue [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+
+    sliderWrapper.value.setAttribute('aria-valuetext', valuetext)
+
+    // label screen reader
+    sliderWrapper.value.setAttribute('aria-label', props.label ? props.label : `slider between ${props.min} and ${props.max}`)
+
+    nextTick().then(() => {
+      resetSize()
+    })
+    window.addEventListener('resize', resetSize)
```
Comment (JeremyWuuuuu) [inline packages/slider/src/index.vue:408]:
```suggestion
on(window, 'resize', resetSize)
```

File: packages/slider/src/index.vue [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+    }
+
+    sliderWrapper.value.setAttribute('aria-valuetext', valuetext)
+
+    // label screen reader
+    sliderWrapper.value.setAttribute('aria-label', props.label ? props.label : `slider between ${props.min} and ${props.max}`)
+
+    nextTick().then(() => {
+      resetSize()
+    })
```
Comment (JeremyWuuuuu) [inline packages/slider/src/index.vue:407]: Use `Async/Await`

File: packages/slider/src/index.vue [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+    window.addEventListener('resize', resetSize)
+  })
+
+  onBeforeUnmount(() => {
+    window.removeEventListener('resize', resetSize)
```
Comment (JeremyWuuuuu) [inline packages/slider/src/index.vue:412]:
```suggestion
off(window, 'resize', resetSize)
```

**Review by jw-foss — CHANGES_REQUESTED** [review]

File: packages/slider/src/useSlide.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
+import { computed, inject, nextTick, ref,Ref, ComputedRef } from 'vue'
+import { ISliderButton, ISliderInitData, ISliderProps, Slide } from './Slider'
+
+export const useSlide = (props:ISliderProps, initData: ISliderInitData, emit):Slide => {
+  const elForm:{ disabled?: boolean;} = inject('elForm', {})
+
+  const slider = ref(null)
+
+  const button1 = ref(null)
```
Comment (zouhangwithsweet) [inline packages/slider/src/useSlide.ts:9]: you should make sense of these vars. Donot use `button1` or `button2`, it's  confusing.

File: packages/slider/src/useStops.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    if (!props.showStops || props.min > props.max) return []
+    if (props.step === 0) {
+      process.env.NODE_ENV !== 'production' &&
+      console.warn('[Element Warn][Slider]step should not be 0.')
+      return []
+    }
+    const stopCount = (props.max - props.min) / props.step
+    const stepWidth = 100 * props.step / (props.max - props.min)
+    const result = []
+    for (let i = 1; i < stopCount; i++) {
```
Comment (zouhangwithsweet) [inline packages/slider/src/useStops.ts:15]: why not use  `Array.prototype.map`

**Review by zouhangwithsweet — CHANGES_REQUESTED** [review]

File: packages/slider/src/index.vue [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+    },
+    marks: {
+      type:Object,
+      default: value => value,
+    },
+  },
+
+  emits: ['update:modelValue', 'change'],
+
+  setup(props: ISliderProps, { emit }){
```
[the remark this answers is not in the document]
Comment (Ryan2128) [reply packages/slider/src/index.vue:195]: To avoid parse error.

**Comment by Ryan2128** [comment]
@zouhangwithsweet  Hi, can you have time re-review this PR? JeremyWuuuuu has no time these days. All requrested changes have been done.

File: packages/slider/src/button.vue [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    @mouseenter="handleMouseEnter"
+    @mouseleave="handleMouseLeave"
+    @mousedown="onButtonDown"
+    @touchstart="onButtonDown"
+    @focus="handleMouseEnter"
+    @blur="handleMouseLeave"
+    @keydown.left="onLeftKeyDown"
+    @keydown.right="onRightKeyDown"
+    @keydown.down.prevent="onLeftKeyDown"
+    @keydown.up.prevent="onRightKeyDown"
```
Comment (JeremyWuuuuu) [inline packages/slider/src/button.vue:17]: Questions why is there only `down` and `up` event gets the `prevent` modifier?

**Review by jw-foss — CHANGES_REQUESTED** [review]

File: packages/slider/src/button.vue [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+      :popper-class="tooltipClass"
+      :disabled="!showTooltip"
+      manual-mode
+      :value="tooltipVisible"
+    >
+      <template #default>{{ formatValue }}</template>
+      <template #trigger>
+        <div class="el-slider__button el-tooltip" :class="{ 'hover': hovering, 'dragging': dragging }"></div>
+      </template>
+    </el-popper>
```
[the remark this answers is not in the document]
Comment (Ryan2128) [reply packages/slider/src/button.vue:31]: YES, I'll use `v-if` instead of `disabled`.

File: packages/slider/src/button.vue (hunk 9 above)
Comment (Ryan2128) [reply packages/slider/src/button.vue:17]: Because when use `tab` key focus the button and press `down` and `up`, the scroll bar also will change.

**Review by jw-foss — APPROVED** [review]

File: packages/slider/src/button.vue (hunk 9 above)
Comment (JeremyWuuuuu) [reply packages/slider/src/button.vue:17]: What about the  X Axis overflows? Have you tried it?

Comment (Ryan2128) [reply packages/slider/src/button.vue:17]: The logic will control button's position on X Axis. I has tried it.

**Pull request closed without merging**